### PR TITLE
Implementation added

### PR DESCRIPTION
**Pull request by McToluene**

### CODE REVIEW

File: src/main/java/com/mctoluene/rental/dtos/request/VideoDetail.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package com.mctoluene.rental.dtos.request;
+
+public class VideoDetail {
```
Comment (vbabxy) [inline src/main/java/com/mctoluene/rental/dtos/request/VideoDetail.java:6]: You can explore lombok... Check out lombok on the interrnet

File: src/main/java/com/mctoluene/rental/services/VideoServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        VideoType videoType = video.getVideoType();
+        double price = videoType.getRate() * days;
+        String type = videoType.getType();
+
+        if (type.equals("Children’s Movie"))
+            price += (double) video.getMaximumAge() / 2;
+        if (type.equals("New Release"))
+            price -= video.getYear();
+        return  price;
+    }
```
Comment (vbabxy) [inline src/main/java/com/mctoluene/rental/services/VideoServiceImpl.java]: Looks good, I need you to take a look at Junit testing

Comment (OjTolu) [reply src/main/java/com/mctoluene/rental/services/VideoServiceImpl.java]: Okay

File: src/main/java/com/mctoluene/rental/services/VideoServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Override
+    public VideoPrice getPrice(VideoDetail detail) {
+        VideoPrice videoPrice = new VideoPrice();
+        try {
+            Video video = videoRepository.findFirstByTitle(detail.getTitle());
+            if (video != null){
+                BeanUtils.copyProperties(video, videoPrice);
+                videoPrice.setPrice(calculatePrice(video, detail.getDays()));
```
Comment (Benjineer) [inline src/main/java/com/mctoluene/rental/services/VideoServiceImpl.java]: see if you could use factory pattern to calculate the price for each type

File: src/main/java/com/mctoluene/rental/repositories/VideoTypeRepository.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.mctoluene.rental.repositories;
+
+import com.mctoluene.rental.entities.VideoType;
+import org.springframework.data.repository.CrudRepository;
+
+public interface VideoTypeRepository extends CrudRepository<VideoType, Long> {
```
Comment (Benjineer) [inline src/main/java/com/mctoluene/rental/repositories/VideoTypeRepository.java]: Just use JpaRepository instead

File: src/main/java/com/mctoluene/rental/repositories/VideoRepository.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.mctoluene.rental.repositories;
+
+import com.mctoluene.rental.entities.Video;
+import org.springframework.data.repository.PagingAndSortingRepository;
+
+public interface VideoRepository extends PagingAndSortingRepository<Video, Long> {
```
Comment (Benjineer) [inline src/main/java/com/mctoluene/rental/repositories/VideoRepository.java]: same here. Just extend JpaRepository instead. It extends this also.

File: src/main/java/com/mctoluene/rental/entities/VideoType.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    public VideoType() {
+    }
+
+    public VideoType(String type, int rate) {
+        this.type = type;
+        this.rate = rate;
+    }
+
+    public Long getId() {
```
Comment (Benjineer) [inline src/main/java/com/mctoluene/rental/entities/VideoType.java]: Research using Lombok library to implement your getters/setters and constructors with simple annotations
@Getter
@Setter
@NoArgumentConstructor

File: src/main/java/com/mctoluene/rental/entities/Video.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
+package com.mctoluene.rental.entities;
+
+import javax.persistence.*;
+
+@Entity
+public class Video {
+
+    @Id
+    @GeneratedValue
```
Comment (Benjineer) [inline src/main/java/com/mctoluene/rental/entities/Video.java]: use @GeneratedValue(strategy = Identity or auto)

File: src/main/java/com/mctoluene/rental/entities/VideoType.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+@Entity
+public class VideoType {
+    @Id
+    @GeneratedValue
+    private Long id;
+
+    @Column(nullable = false)
+    private String type;
+
+    private int rate;
```
Comment (Benjineer) [inline src/main/java/com/mctoluene/rental/entities/VideoType.java]: Use Object wrapper Integer instead

File: src/main/java/com/mctoluene/rental/dtos/response/VideoPrice.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.mctoluene.rental.dtos.response;
+
+public class VideoPrice {
+    private double price;
```
Comment (Benjineer) [inline src/main/java/com/mctoluene/rental/dtos/response/VideoPrice.java]: Object wrapper Double instead

File: src/main/java/com/mctoluene/rental/dtos/response/PagedResponse.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
+package com.mctoluene.rental.dtos.response;
+
+public class PagedResponse<T> extends Response<T> {
```
Comment (Benjineer) [inline src/main/java/com/mctoluene/rental/dtos/response/PagedResponse.java]: Nice work here. Research using spring Page, Pageable to implement pagination

File: src/main/java/com/mctoluene/rental/controllers/VideoController.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    @GetMapping
+    public PagedResponse get(@RequestParam int pageNumber, @RequestParam int pageSize){
+        PaginationFilter filter = new PaginationFilter(pageNumber, pageSize);
+        var videos = videoService.getVideos(filter);
+        return new PagedResponse<>(videos.getContent(), "Videos fetched successfully", true, filter.getPageNumber(), filter.getPageSize());
+    }
+
+    @GetMapping("/price")
```
Comment (Benjineer) [inline src/main/java/com/mctoluene/rental/controllers/VideoController.java]: was thinking of something like a method called calculatePrice POST _**video/{id}/price**_ body:{"days": 1} or POST **_video/{title}/price_** body:{"days": 1} . Looks more Restful to me. RequestParam/QueryParams are usually for optional params/filters/sort

File: src/main/java/com/mctoluene/rental/controllers/VideoController.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+@RequestMapping("video")
+public class VideoController {
+    private final VideoService videoService;
+
+    public VideoController(VideoService videoService) {
+        this.videoService = videoService;
+    }
+
+    @GetMapping
+    public PagedResponse get(@RequestParam int pageNumber, @RequestParam int pageSize){
```
Comment (Benjineer) [inline src/main/java/com/mctoluene/rental/controllers/VideoController.java]: yeah the thing about spring Page and Pageable for pagination

**Review by Benjineer — CHANGES_REQUESTED** [review]
Excellent job overall. Just try to implement some of the suggestions and add tests for all layers(Controller,Service and Repository)

**Pull request closed without merging**